File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/java/sparse-table.java
/**
 * Java Implementation: Sparse Table (Range Minimum Query)
 * Invariant: Precomputation O(N log N), Range Query O(1)
 */
package seed.dsa.core;

import java.util.*;

public class SparseTableRangeMinimumQuery<T> {
    private final List<T> items = new ArrayList<>();

    // TODO: Implement your solution here
    public void add(T value) {
        items.add(value);
    }

    public int size() {
        return items.size();
    }
}

File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/cpp/sparse-table.cpp
/**
 * C++ Implementation: Sparse Table (Range Minimum Query)
 * Invariant: Precomputation O(N log N), Range Query O(1)
 */
#include <iostream>
#include <vector>
#include <stdexcept>

template <typename T>
class SparseTableRangeMinimumQuery {
private:
    std::vector<T> elements;
public:
    // TODO: Implement primary invariant operations
    void push(const T& val) {
        elements.push_back(val);
    }
    size_t size() const { return elements.size(); }
};
